libtenzir/builtins/operators/load_tcp.cpp

```diff
@@ -0,0 +1,864 @@
+  }
+
+  auto operator()(generator<Elements> input, operator_control_plane& ctrl) const
+    -> generator<std::monostate> {
+    auto connection_manager = connection_manager_.lock();
+    TENZIR_ASSERT(connection_manager);
+    for (auto&& elements : input) {
+      ctrl.set_waiting(true);
+      ctrl.self()
+        .request(connection_manager, caf::infinite, atom::write_v,
+                 std::move(elements))
+        .then(
+          [&]() {
+            ctrl.set_waiting(false);
+          },
+          [&](const caf::error& err) {
+            diagnostic::error(err)
+              .note("failed to read from TCP connection")
+              .emit(ctrl.diagnostics());
+          });
+      co_yield {};
+    }
+  }
+
+  auto name() const -> std::string override {
+    return fmt::format("internal-load-tcp-sink-{}",
+                       operator_type_name<Elements>());
+  }
+
+  auto optimize(const expression& filter, event_order order) const
+    -> optimize_result override {
+    TENZIR_UNUSED(filter, order);
+    return do_not_optimize(*this);
+  }
+
+  auto infer_type_impl(operator_type input) const
+    -> caf::expected<operator_type> override {
+    TENZIR_ASSERT(input.is<Elements>());
+    return tag_v<void>;
+  }
+
+  friend auto inspect(auto& f, load_tcp_sink_operator& x) -> bool {
+    return f.apply(x.connection_manager_);
+  }
+
+private:
+  detail::weak_handle<connection_manager_actor<Elements>> connection_manager_
+    = {};
+};
+
+// -- connection-manager actor -------------------------------------------------
+
+template <class Elements>
+struct connection_manager_state {
+  [[maybe_unused]] static constexpr auto name = "connection-manager";
+
+  connection_manager_actor<Elements>::pointer self = {};
+  load_tcp_args args = {};
+  shared_diagnostic_handler diagnostics = {};
+
+  // Everything required for the I/O worker.
+  std::vector<std::thread> io_workers = {};
+  std::shared_ptr<boost::asio::io_context> io_ctx = {};
+  std::optional<boost::asio::ip::tcp::socket> socket = {};
+
+  // Everything required for listening for connections.
+  std::optional<boost::asio::ip::tcp::acceptor> acceptor = {};
+
+  // Everything needed for back pressure handling.
+  static constexpr auto max_buffered_batches = size_t{20};
+  std::queue<Elements> buffer = {};
+  caf::typed_response_promise<Elements> read_rp = {};
+  std::queue<caf::typed_response_promise<void>> write_rps = {};
+
+  // State we need to keep for each peer.
+  struct connection_state {
+    static constexpr auto max_queued_chunks = size_t{10};
+    static constexpr auto read_buffer_size = size_t{65'536};
+
+    std::optional<boost::asio::ip::tcp::socket> socket = {};
+    std::optional<boost::asio::ssl::context> ssl_ctx = {};
+    std::optional<boost::asio::ssl::stream<boost::asio::ip::tcp::socket&>>
+      tls_socket = {};
+    pipeline_executor_actor pipeline_executor = {};
+    std::vector<char> read_buffer = {};
+    mutable std::mutex mutex = {};
```

Comment:
We should add a comment here explaining which parts this locks and why and why the remaining parts don't need to be locked.